File: markdown/group/group_publications.py
import argparse
import html
import json
from collections.abc import Mapping, Sequence
from typing import Any

from markdown.utils.metadata import NOTES, PUBLICATIONS, TITLE, WORDCOUNT
from markdown.utils.publications import Publication, Publications, State


def generate_header(venues: Sequence[str]) -> list[str]:
    return [
        "<thead>",
        "<tr>",
        '<th title="Target">Target</th>',
        '<th title="Title">Title</th>',
        '<th title="Wordcount">Wordcount</th>',
        '<th style="border-right: 3px solid" title="Notes">Notes</th>',
        *[f'<th title="{html.escape(v)}">{html.escape(v, quote=False)}</th>' for v in venues],
        "</tr>",
        "</thead>",
    ]


def generate_row(
    target: str,
    data: Publications,
    venues: Sequence[str],
    raw: Mapping[str, Any],
) -> list[str]:
    ps = {}
    for p in data.publications:
        ps[p.venue] = p

    title = raw.get(TITLE, "")
    wordcount = raw.get(WORDCOUNT, "")
    notes = raw.get(NOTES, "")

    class_attr = ""
    if data.highest_active_state:
        class_attr = data.highest_active_state.name.lower()

    out = [
        "<tr>",
        f'<td class="{class_attr}" title="{html.escape(target)}">'
        f'<a href="#{html.escape(target)}">{html.escape(target, quote=False)}</a></td>',
        f'<td title="{html.escape(title)}">{html.escape(title, quote=False)}</td>',
        f'<td title="{html.escape(wordcount)}">{html.escape(wordcount, quote=False)}</td>',
        f'<td style="border-right: 3px solid" title="{html.escape(notes)}">'
        f"{html.escape(notes, quote=False)}</td>",
    ]

    for v in sorted(venues):
        if v in ps:
            out.append(generate_cell(target, ps[v]))
        else:
            out.append("<td></td>")

    out.append("</tr>")
    return out


def generate_cell(target: str, p: Publication) -> str:
    content = [
        d.date.isoformat() + " " + d.state.name.lower().replace("_", "-").capitalize()
        for d in p.dates
    ]
    return (
        f'<td class="{p.latest.state.name.lower()}" title="{html.escape(target + ", " + p.venue)}">'
        f'<a href="#{html.escape(target)}">'
        f'{"<br>".join([html.escape(c, quote=False) for c in content])}</a></td>'
    )


def generate_table(data: Mapping[str, Publications], raw: Mapping[str, Any]) -> list[str]:
    out = ["<table>"]

    venue_set = set()
    for ps in data.values():
        for p in ps.publications:
            venue_set.add(p.venue)
    venues = sorted(venue_set)

    out += generate_header(venues)

    out.append("<tbody>")
    for target, ps in sorted(data.items()):
        out += generate_row(target, ps, venues, raw[target])
    out += ["</tbody>", "</table>"]

    return out


def generate_details(raw: Mapping[str, Any]) -> list[str]:
    out = ["<h2>Details</h2>"]
    for target in sorted(raw):
        if PUBLICATIONS in raw[target] and raw[target][PUBLICATIONS]:
            out += [
                f'<h3 id="{html.escape(target)}">{html.escape(target, quote=False)}</h3>',
                "<code><pre>{}</pre></code>".format(  # noqa: UP032
                    html.escape(json.dumps(raw[target], sort_keys=True, indent=4), quote=False),
                ),
            ]
    return out


def generate_head() -> list[str]:
    return [
        "<head>",
        '<meta charset="utf-8">',
        "<title>Publications</title>",
        "<style>",
        "table { border-collapse: collapse; }",
        "th, td { border: 1px solid; padding: 5px; }",
        "a:link { color: black; }",
        "a:visited { color: black; }",
        f".{State.SUBMITTED.name.lower()} {{ background-color: #ffff00; }}",
        f".{State.REJECTED.name.lower()} {{ background-color: #ff6d6d; }}",
        f".{State.WITHDRAWN.name.lower()} {{ background-color: #ff972f; }}",
        f".{State.ABANDONED.name.lower()} {{ background-color: #cccccc; }}",
        f".{State.ACCEPTED.name.lower()} {{ background-color: #729fcf; }}",
        f".{State.SELF_PUBLISHED.name.lower()} {{ background-color: #158466; }}",
        f".{State.PUBLISHED.name.lower()} {{ background-color: #81d41a; }}",
        "</style>",
        "</head>",
    ]


def generate_body(data: Mapping[str, Publications], raw: Mapping[str, Any]) -> list[str]:
    out = [
        "<body>",
        "<h1>Publications</h1>",
    ]
    out += generate_table(data, raw)
    out += generate_details(raw)
    out += ["</body>"]
    return out


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("metadata_file")
    parser.add_argument("out_file")
    args = parser.parse_args()

    with open(args.metadata_file, encoding="utf-8") as f:
        j = json.load(f)

    data = {
        k: Publications.model_validate_json(json.dumps(v[PUBLICATIONS]))
        for k, v in j.items()
        if PUBLICATIONS in v
    }

    out = [
        "<!doctype html>",
        '<html lang="en-GB">',
    ]

    out += generate_head()
    out += generate_body(data, j)

    out += [
        "</html>",
    ]

    with open(args.out_file, "w", encoding="utf-8") as outfile:
        outfile.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()

File: markdown/private/group/group_publications.rs
use clap::Parser;
use markdown::args;
use markdown::json::{JsonSerializable, from_json};
use markdown::metadata::{MetadataMap, OutputMetadata};
use markdown::publications::{Publication, State};
use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::error::Error;
use std::fs::{read_to_string, write};

fn capitalise(s: &str) -> String {
    let mut chars: Vec<char> = s.chars().collect();
    if let Some(c) = chars.first_mut() {
        c.make_ascii_uppercase();
    }
    String::from_iter(chars)
}

#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Cli {
    #[arg(value_parser = args::non_empty())]
    metadata_file: String,

    #[arg(value_parser = args::non_empty())]
    out_file: String,
}

fn generate_header(venues: &BTreeSet<&str>) -> Vec<String> {
    let mut out = vec![
        String::from("<thead>"),
        String::from("<tr>"),
        String::from("<th title=\"Target\">Target</th>"),
        String::from("<th title=\"Title\">Title</th>"),
        String::from("<th title=\"Wordcount\">Wordcount</th>"),
        String::from("<th style=\"border-right: 3px solid\" title=\"Notes\">Notes</th>"),
    ];
    out.extend(venues.iter().map(|v| {
        format!(
            "<th title=\"{}\">{}</th>",
            html_escape::encode_double_quoted_attribute(v),
            html_escape::encode_text(v)
        )
    }));
    out.extend([String::from("</tr>"), String::from("</thead>")]);
    out
}

fn generate_row(target: &str, venues: &BTreeSet<&str>, metadata: &OutputMetadata) -> Vec<String> {
    let ps: HashMap<&str, &Publication> = HashMap::from_iter(
        metadata
            .publications()
            .publications()
            .iter()
            .map(|p| (p.venue(), p)),
    );

    let class_attr = {
        if let Some(state) = metadata.publications().highest_active_state() {
            state.to_string()
        } else {
            String::new()
        }
    };

    let mut out = vec![
        String::from("<tr>"),
        format!(
            "<td class=\"{class_attr}\" title=\"{}\"><a href=\"#{}\">{}</a></td>",
            html_escape::encode_double_quoted_attribute(target),
            html_escape::encode_double_quoted_attribute(target),
            html_escape::encode_text(target)
        ),
        format!(
            "<td title=\"{}\">{}</td>",
            html_escape::encode_double_quoted_attribute(metadata.title().unwrap_or(&String::new())),
            html_escape::encode_text(metadata.title().unwrap_or(&String::new())),
        ),
        format!(
            "<td title=\"{}\">{}</td>",
            html_escape::encode_double_quoted_attribute(&metadata.wordcount().to_string()),
            html_escape::encode_text(&metadata.wordcount().to_string())
        ),
        format!(
            "<td style=\"border-right: 3px solid\" title=\"{}\">{}</td>",
            html_escape::encode_double_quoted_attribute(metadata.notes().unwrap_or(&String::new())),
            html_escape::encode_text(metadata.notes().unwrap_or(&String::new()))
        ),
    ];

    for v in venues {
        out.push(if let Some(p) = ps.get(v) {
            generate_cell(target, p)
        } else {
            String::from("<td></td>")
        });
    }

    out.push(String::from("</tr>"));
    out
}

fn generate_cell(target: &str, p: &Publication) -> String {
    let content = Vec::from_iter(p.dates().iter().map(|d| {
        format!(
            "{} {}",
            d.date.format("%Y-%m-%d"),
            capitalise(&d.state.to_string().replace("_", "-"))
        )
    }));

    format!(
        "<td class=\"{}\" title=\"{}\"><a href=\"#{}\">{}</a></td>",
        p.latest().state,
        html_escape::encode_double_quoted_attribute(&format!("{target}, {}", p.venue())),
        html_escape::encode_double_quoted_attribute(target),
        Vec::from_iter(content.iter().map(html_escape::encode_text)).join("<br>")
    )
}

fn generate_table(metadata: &MetadataMap) -> Vec<String> {
    let mut out = vec![String::from("<table>")];

    let mut venues = BTreeSet::new();
    for m in metadata.data().values() {
        for p in m.publications().publications() {
            venues.insert(p.venue());
        }
    }

    out.extend(generate_header(&venues));

    out.push(String::from("<tbody>"));
    for (target, m) in metadata.data() {
        out.extend(generate_row(target, &venues, m));
    }
    out.extend([String::from("</tbody>"), String::from("</table>")]);

    out
}

fn generate_details(metadata: &MetadataMap) -> Result<Vec<String>, Box<dyn Error>> {
    let mut out = vec![String::from("<h2>Details</h2>")];
    for (target, m) in metadata.data() {
        if !m.publications().is_empty() {
            out.extend([
                format!(
                    "<h3 id=\"{}\">{}</h3>",
                    html_escape::encode_double_quoted_attribute(target),
                    html_escape::encode_text(target)
                ),
                format!(
                    "<code><pre>{}</pre></code>",
                    html_escape::encode_text(&m.to_json()?)
                ),
            ]);
        }
    }
    Ok(out)
}

fn generate_head() -> Vec<String> {
    vec![
        String::from("<head>"),
        String::from("<meta charset=\"utf-8\">"),
        String::from("<title>Publications</title>"),
        String::from("<style>"),
        String::from("table { border-collapse: collapse; }"),
        String::from("th, td { border: 1px solid; padding: 5px; }"),
        String::from("a:link { color: black; }"),
        String::from("a:visited { color: black; }"),
        format!(
            ".{} {{ background-color: #ffff00; }}",
            State::Submitted.to_string()
        ),
        format!(
            ".{} {{ background-color: #ff6d6d; }}",
            State::Rejected.to_string()
        ),
        format!(
            ".{} {{ background-color: #ff972f; }}",
            State::Withdrawn.to_string()
        ),
        format!(
            ".{} {{ background-color: #cccccc; }}",
            State::Abandoned.to_string()
        ),
        format!(
            ".{} {{ background-color: #729fcf; }}",
            State::Accepted.to_string()
        ),
        format!(
            ".{} {{ background-color: #158466; }}",
            State::SelfPublished.to_string()
        ),
        format!(
            ".{} {{ background-color: #81d41a; }}",
            State::Published.to_string()
        ),
        String::from("</style>"),
        String::from("</head>"),
    ]
}

fn generate_body(metadata: &MetadataMap) -> Result<Vec<String>, Box<dyn Error>> {
    let mut out = vec![
        String::from("<body>"),
        String::from("<h1>Publications</h1>"),
    ];
    out.extend(generate_table(metadata));
    out.extend(generate_details(metadata)?);
    out.push(String::from("</body>"));
    Ok(out)
}

fn main() -> Result<(), Box<dyn Error>> {
    let args = Cli::parse();

    let metadata = MetadataMap::build(BTreeMap::from_iter(
        from_json::<MetadataMap>(&read_to_string(args.metadata_file)?)?
            .data()
            .iter()
            .filter(|(_, m)| !m.publications().is_empty())
            .map(|(target, m)| (target.clone(), m.clone())),
    ))?;

    let mut out = vec![
        String::from("<!doctype html>"),
        String::from("<html lang=\"en-GB\">"),
    ];

    out.extend(generate_head());
    out.extend(generate_body(&metadata)?);

    out.push(String::from("</html>"));

    write(args.out_file, out.join("\n") + "\n")?;
    Ok(())
}
